### PR TITLE
[BREAKING] Add new reserved keywords.

### LINKED ISSUE #2182: Reserve additional keywords
`const`, `immutable` (#715), `mutable`, `promise`

### PR DESCRIPTION
**Pull request by ekpyron**
Closes #2182.

### CODE REVIEW

**Comment by chriseth** [comment]
If you add `Closes #2182` to the description of the pull request, that issue will be automatically closed and also linked to this pull request.

File: test/libsolidity/SolidityParser.cpp [hunk 1]
```diff
@@ -1164,6 +1164,56 @@ BOOST_AUTO_TEST_CASE(constant_is_keyword)
...
+		"topic",
+		"try",
+		"type",
+		"typedef",
+		"typeof",
+		"unchecked"
+	};
+	
+
+	for (const auto& keyword : keywords) {
```
Comment (chriseth) [inline test/libsolidity/SolidityParser.cpp]: Coding style: no space in front of `:` except in `a ? b : c`.

Comment (chriseth) [reply test/libsolidity/SolidityParser.cpp]: And `{` on a line of its own.

**Comment by chriseth** [comment]
Please also add a line to `Changelog.txt` describing the change. It is probably a good idea for this PR to not add it to the changelog of 0.4.21 but instead start a new version 0.5.0. We will then merge it when 0.5.0 will be the next release.

File: test/libsolidity/SolidityParser.cpp (hunk 1 above)
Comment (erak) [reply test/libsolidity/SolidityParser.cpp]: Thanks for pointing that out :)

**Comment by chriseth** [comment]
Looks good! I'm not adding a positive review on purpose, so we do not accidentally merge this PR.

File: test/libsolidity/SolidityParser.cpp [hunk 2]
```diff
@@ -1164,6 +1164,56 @@ BOOST_AUTO_TEST_CASE(constant_is_keyword)
 	CHECK_PARSE_ERROR(text, "Expected identifier");
 }
 
+BOOST_AUTO_TEST_CASE(keyword_is_reserved)
```
Comment (axic) [inline test/libsolidity/SolidityParser.cpp]: Can this test case be extracted and merged with current keywords?

**Review by axic — DISMISSED** [review]
There is a section in the documentation listing the reserved keywords, that should be updated too.

File: test/libsolidity/SolidityParser.cpp (hunk 2 above)
Comment (ekpyron) [reply test/libsolidity/SolidityParser.cpp]: OK, see #3633

**Comment by axic** [comment]
Please rebase since #3633 was merged.

**Comment by chriseth** [comment]
We should also add `copy` and `copyof`.

**Comment by axic** [comment]
> If you add `Closes #2182` to the description of the pull request

Please do that and do not include it in the commit messages.

**Comment by ekpyron** [comment]
I removed the ``constructor`` keyword from this PR, since it breaks the tests unless the parser is updated, so it makes more sense to only add it in #4106.

File: Changelog.md [hunk 3]
```diff
@@ -4,6 +4,9 @@ Breaking Changes:
  * Type Checker: Disallow arithmetic operations for Boolean variables.
 
 Features:
+ * Adds new reserved keywords: ``alias``, ``apply``, ``auto``, ``const``, ``copy``, ``copyof``, ``define``, ``fallback``,
```
Comment (chriseth) [inline Changelog.md]: `copy` could be rather widely used. Do we really want to reserve that?

**Comment by chriseth** [comment]
Do we really want to reserve all of them? Perhaps we should rather make breaking changes more often...

**Comment by ekpyron** [comment]
@chriseth I find the list of new keywords (and especially ``copy``) to be a bit excessive myself and it seems to me more frequent breaking releases make more sense.
The question is whether there is a subset of this list that should definitely be in 0.5.0.

**Comment by chriseth** [comment]
100%: constructor and emit
80%: mutable, immutable, unchecked, override
60%: alias, typedef

**Comment by axic** [comment]
Rebased.

**Comment by ekpyron** [comment]
@axic We still need to decide whether we want to keep this PR at all or which subset to keep.
``constructor`` and ``emit`` are handled in separate commits, but we need to decide which of the other ones should in fact be reserved in 0.5.0 already.

**Comment by axic** [comment]
I think we should just merge this. The only debatable one I think is `topic` and perhaps `fallback` because we don't know yet if we'll have `wildcard` or `fallback`, etc.

**Comment by ekpyron** [comment]
@axic I think what started the discussion actually was whether to keep ``copy``, since it may be commonly used and break a lot of code. (@chriseth brought that up in an now outdated Changelog.md review comment)

**Comment by axic** [comment]
Also if this is merged first and `constructor`/`emit` later then those are just moved from the reserved list to the "new keywords" line.

**Comment by axic** [comment]
It seems `copy` is used in stringutils so that might be something we need to consider. Nothing else seems to break.

**Comment by axic** [comment]
I'd take a shortcut and merge the non-contentious ones here and open a new PR for the rest.

**Comment by ekpyron** [comment]
Removed ``copy``, ``topic`` and ``fallback`` from the list.
I would still not include ``constructor`` and ``emit`` here, since the workarounds will fail once the keywords are introduced, so I think it's best to handle them separately altogether.

**Review by axic — DISMISSED** [review]

**Comment by axic** [comment]
Confirmed against the issue.

**Comment by ekpyron** [comment]
Removed ``const`` as well.

**Review by axic — APPROVED** [review]

**Pull request merged**